Fix save_to_file to be indentation independent

ShimShtein opened the pull request:
Heredoc notation is very sensitive to leading spaces.
This implementation makes the `save_to_file` macro safe for indentation.

ShimShtein committed:
Fixes #37741 - Fix save_to_file to be indentation independent

ekohl commented on app/services/foreman/renderer/scope/macros/base.rb:
```diff
@@ -124,8 +124,12 @@ def save_to_file(filename, content, verbatim: false)
-              content += "\n" unless content.end_with?("\n")
-              "cat << #{delimiter} > #{filename}\n#{content}#{delimiter}"
+              content_echos = content.split("\n").map do |content_line|
+                "echo #{Shellwords.shellescape(content_line)} >> #{filename}"
```
Previously the code allowed to use variables like `$var` but now that's going to be escaped, making it backwards incompatible. If you want it really verbatim, we have the base64 method above.

ShimShtein replied:
This should be possible now

ShimShtein replied:
Now the `verbatim` will escape everything and the `else` will escape only double quotes.

ekohl replied:
@adamruzicka you initially introduced this because you found a case where it broke. Do you think this is sufficient?

adamruzicka replied:
I don't really see where this wouldn't work, but I don't have a good feeling about it.

As a counter-proposal, how about we keep base64, but on the ruby side we drop the newlines from it and then `echo` it? Something like 
```ruby
content = Base64.encode64(content).tr("\n", "")
"echo #{content} | base64 -d > #{filename}
```

ShimShtein replied:
The line can get quite long. Is there any limit that we may hit with too long lines?

adamruzicka replied:
There doesn't seem to be, although shell might need a lot of time and memory to parse the script once we get into the hundreds of megabytes range. As far as I know, all the inputs passed into save_by_file are fairly small, right?

ShimShtein replied:
@adamruzicka https://serverfault.com/questions/163371/linux-command-line-character-limit says that there is some limit.

I agree that it's quite long and we probably won't hit it soon, but I preferred not to rely on any of it.
I have tried to work with base64, but somehow I wasn't able to make it work properly with multiple lines and variable name. I can use a tempfile, but it seems a bit of overkill.

adamruzicka replied:
> says that there is some limit.

Well, yes and no. From what I've seen, what's written there applies when launching things that are external to the shell, but it doesn't seem to apply to shell builtins, such as echo in this case.

> I have tried to work with base64, but somehow I wasn't able to make it work properly with multiple lines and variable name

How about we mix the two approaches?

```
b64 = Base64.encode64(content)
content_echos = b64.each_line.map { |l| "echo #{l}" }.join
"(#{content_echos}) | base64 -d >#{filename}"
```

ShimShtein replied:
If you go the `echo` path, you need to accumulate the result somewhere. and once you have the whole base64 string accumulated, you can pass it into `base64`. 
Let me try to accumulate the base64 value without newlines and then pipe it into base64.
Something like:
```
b64 = Base64.encode64(content)
content_var = b64.each_line.map { |l| "FOO+=#{l}" }.join("\n")
content_var + "echo $FOO | base64 -d >#{filename}"
```
Mostly I had trouble with the line endings. Maybe in this approach it will be a bit easier.

But let's say this works, we will have different implementations for `verbatim` and `else` use cases.
On the other hand, if we go with the escaping approach, we will be able to DRY the code by extracting most of the generation into a single method, and just changing the escaping method of the input.

adamruzicka replied:
> If you go the echo path, you need to accumulate the result somewhere

Well, you can always spawn a subshell just like I did in the previous suggestion.

> On the other hand, if we go with the escaping approach, we will be able to DRY the code by extracting most of the generation into a single method, and just changing the escaping method of the input.

True, at the same time we lose the option to write out binary blobs

ShimShtein replied:
Reimplemented as base64 single string.

ekohl commented on app/services/foreman/renderer/scope/macros/base.rb:
```diff
@@ -124,8 +124,12 @@ def save_to_file(filename, content, verbatim: false)
               content = Base64.encode64(content)
               "cat << #{delimiter} | base64 -d > #{filename}\n#{content}#{delimiter}"
```
This case will still have an issue with `indent()`.

ShimShtein replied:
Do you think we will have `\n` symbols there? I can implement the same solution here too

ShimShtein replied:
Changed the verbatim option to stop using heredoc too

ekohl reviewed (changes requested):
I don't think you're going to find any solution that works together with indent.

ShimShtein commented:
This should fix the indentation error mentioned in https://github.com/theforeman/foreman/pull/10283

ekohl reviewed (changes requested):
This is not a change I feel comfortable cherry picking so I'd still prefer to revert the offending change in the short term and then figure this out.

ShimShtein commented:
but then we will be left with broken cloud-init, unless we redesign it as you suggest. Additionally we will be left with `save_to_file` that is unusable if indented

ShimShtein commented:
So I tracked down the whole chain of events:
* https://github.com/theforeman/foreman/pull/10208 introduced a bug into cloud-init (it made the code indented incorrectly, so cloud-init yaml bcame invalid)
* https://github.com/theforeman/foreman/pull/10153 fixed the indentation problem, but exposed a lingering problem with `save_to_file`, since this helper method does not work well with indentation

Now this PR tries to avoid reverting the previous two

ekohl commented:
>     * [Fixes #37433 - Set indentation when calling subscription_manager_setup #10200](https://github.com/theforeman/foreman/pull/10200) fixed the indentation problem, but exposed a lingering problem with `save_to_file`, since this helper method does not work well with indentation

I think you mean https://github.com/theforeman/foreman/pull/10153

ShimShtein commented:
Updated the original comment

ekohl commented on test/static_fixtures/script.txt:
```diff
@@ -0,0 +1,5 @@
+#!/bin/sh
+
+echo "Calling Ansible AWX/Tower provisioning callback..."
+/usr/bin/curl -v -k -s --data "host_config_key=" https:///api/v2/job_templates//callback/
```
This is missing some interpolated values. Can we somehow set up the environment so those are set?

ShimShtein replied:
Added another test with interpolated values.

ShimShtein commented:
I have replaced base64 implementation with `shellescape` one. The whole idea of `verbatim` flag is to place the exact strings in the file as opposed to parameter interpolation in `verbatim=false` use case.

I have implemented them both in a similar manner with the only difference what symbols should be escaped: in verbatim everything is escaped, in non-verbatim only `"` is escaped.

adamruzicka commented on app/services/foreman/renderer/scope/macros/base.rb:
```diff
@@ -117,15 +117,25 @@ def pxe_kernel_options
+              end.join("\n")
+
+              # prefix the append commands with a cleanup command
+              "> #{filename}\n#{content_echos}"
```
This relies on stdin being closed, right? Can we safely assume that stdin will be closed in all the scenarios where the save_to_file macro might be used?

ShimShtein replied:
Thanks! Now the implementation does not rely on stdin.

ShimShtein commented:
@adamruzicka forgot to push the version. Now it should work on all OS-es. What do you think about this implementation?

adamruzicka commented on app/services/foreman/renderer/scope/macros/base.rb:
```diff
@@ -117,15 +117,21 @@ def pxe_kernel_options
             elsif verbatim
-              content = Base64.encode64(content)
-              "cat << #{delimiter} | base64 -d > #{filename}\n#{content}#{delimiter}"
+              base64 = Base64.encode64(content)
+              content_assignments = base64.split("\n").map { |l| "echo #{Shellwords.shellescape(l)}" }.join("\n")
```
Is the escaping even needed? Base64 shouldn't contain any problematic characters.

adamruzicka replied:
@ShimShtein ping

adamruzicka reviewed (commented):
One last question, otherwise lgtm
